src/Service/TrustedDevice/Http/CookieHelper.php

```diff
@@ -0,0 +1,118 @@
+        private readonly LoggerInterface $logger
+    ) {
+    }
+
+    public function write(Response $response, CookieValueInterface $value): void
+    {
+        $cookieName = $this->buildCookieName($value->getUserId(), $value->getNotificationAddress());
+
+        // The CookieValue is encrypted
+        $encryptedCookieValue = $this->encryptionHelper->encrypt($value);
+        $fingerprint = $this->hashFingerprint($encryptedCookieValue);
+        $this->logger->notice(sprintf('Writing a trusted-device cookie with fingerprint %s', $fingerprint));
+        // Create a Symfony HttpFoundation cookie object
+        $cookie = $this->createCookieWithValue($encryptedCookieValue, $cookieName);
+        // Which is added to the response headers
+        $response->headers->setCookie($cookie);
+    }
+
+    /**
+     * Retrieve the current cookie from the Request if it exists.
+     */
+    public function read(Request $request, string $userId, string $notificationAddress): CookieValueInterface
+    {
+        $cookieName = $this->buildCookieName($userId, $notificationAddress);
+        if (!$request->cookies->has($cookieName)) {
+            throw new CookieNotFoundException();
+        }
+        $cookie = $request->cookies->get($cookieName);
+        if (!is_string($cookie)) {
+            throw new InvalidArgumentException('Cookie payload must be string.');
+        }
+        $fingerprint = $this->hashFingerprint($cookie);
+        $this->logger->notice(sprintf('Reading a trusted-device cookie with fingerprint %s', $fingerprint));
+        return $this->encryptionHelper->decrypt($cookie);
+    }
+
+    public function fingerprint(Request $request, string $userId, string $notificationAddress): string
+    {
+        $cookieName = $this->buildCookieName($userId, $notificationAddress);
+        if (!$request->cookies->has($cookieName)) {
+            throw new CookieNotFoundException();
+        }
+        $cookie = $request->cookies->get($cookieName);
+        if (!is_string($cookie)) {
+            throw new InvalidArgumentException('Cookie payload must be string.');
+        }
+        return $this->hashFingerprint($cookie);
+    }
+
+    private function createCookieWithValue(string $value, string $name): Cookie
+    {
+        return new Cookie(
+            $name,
+            $value,
+            $this->getTimestamp($this->configuration->lifetimeInSeconds),
+            '/',
+            null,
+            true,
+            true,
+            false,
+            $this->configuration->sameSite->value
+        );
+    }
+
+    private function hashFingerprint(string $encryptedCookieValue): string
+    {
+        return hash('sha256', $encryptedCookieValue);
+    }
+
+    private function getTimestamp(int $expiresInSeconds): int
+    {
+        $currentTimestamp = time();
+        return $currentTimestamp + $expiresInSeconds;
+    }
+
+    public function buildCookieName(string $userId, string $notificationAddress): string
```

Comment:
This smells like a private method.. I see you only use it in test so it should be possible to make it private

Reply:
:+1: changed